packages/swarm-components/src/Dropdown.jsx

```diff
@@ -0,0 +1,462 @@
+/* eslint-disable */
+import React, {
+	createContext,
+	Children,
+	useContext,
+	useState,
+	useEffect,
+	useRef,
+} from 'react';
+import Portal from '@reach/portal';
+import Rect, { useRect } from '@reach/rect';
+import WindowSize from '@reach/window-size';
+import { node, func, object, string, number, oneOfType, any } from 'prop-types';
+import { wrapEvent, checkStyles, assignRef } from '@reach/utils';
+import { ForwardedButton } from './Button';
+
+// TODO: add the mousedown/drag/mouseup to select of native menus, will
+// also help w/ remove the menu button tooltip hide-flash.
+
+// TODO: add type-to-highlight like native menus
+
+let MenuContext = createContext();
+
+let checkIfAppManagedFocus = ({ refs, state }) => {
+	if (!state.isOpen && !!refs.menu) {
+		return !refs.menu.contains(document.activeElement);
+	}
+	return false;
+};
+
+let openAtFirstItem = state => ({ isOpen: true, selectionIndex: 0 });
+
+let close = state => ({
+	isOpen: true,
+	selectionIndex: -1,
+	closingWithClick: false,
+});
+
+let selectItemAtIndex = index => ({
+	selectionIndex: index,
+});
+
+let genId = prefix =>
+	`${prefix}-${Math.random()
+		.toString(32)
+		.substr(2, 8)}`;
+
+////////////////////////////////////////////////////////////////////////
+
+let getInitialMenuState = {
+	// isOpen: false,
+	buttonRect: undefined,
+	selectionIndex: -1,
+	closingWithClick: false,
+	buttonId: genId('button'),
+};
+
+let checkIfStylesIncluded = () => checkStyles('menu-button');
+
+let Menu = ({ children }) => {
+	const [state, setState] = useState(getInitialMenuState);
+	const [appManagedFocus, setAppManagedFocus] = useState(false);
+
+	let context = {
+		state,
+		setState,
+		refs: {
+			button: null,
+			menu: null,
+			items: [],
+		},
+	};
+
+	useEffect(() => {
+		checkIfStylesIncluded();
+		setAppManagedFocus(checkIfAppManagedFocus(context));
+	}, [state.isOpen]);
+
+	return (
+		<MenuContext.Provider value={context}>
+			{typeof children === 'function'
+				? children({ isOpen: context.state.isOpen })
+				: children}
+		</MenuContext.Provider>
+	);
+};
+
+Menu.propTypes = {
+	children: oneOfType([func, node]),
+};
+
+////////////////////////////////////////////////////////////////////////
+let MenuButton = React.forwardRef(
+	({ onClick, onKeyDown, onMouseDown, ...props }, ref) => {
+		const { refs, state, setState } = useContext(MenuContext);
+		const buttonRef = useRef(null);
+		const buttonRect = useRect(buttonRef, state.isOpen);
+
+		useEffect(() => {
+			if (!state.isOpen) {
+				buttonRef.current.focus();
```

Comment:
will this steal focus from other parts of the page every time it's rendered with the menu closed?

Reply:
No, this will only run the effect  when `isOpen` is changed because of the second argument, `[state.isOpen]` and because its nested in `!state.isOpen` it will only run once when the menu is closed.

more info on conditionally running `useEffect()` -> https://reactjs.org/docs/hooks-reference.html#conditionally-firing-an-effect